pkg/pattern/tokenization/replacer.go

```diff
@@ -0,0 +1,859 @@
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// Linux, e.g.
+		// - "Mon  2 Jan 15:04:05 MST 2006"
+		// - "Tue 23 Jan 15:04:05 -0700 2023"
+		if r.advanceChar(' ') && (r.advanceChar(' ') || true) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceMonthName() && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(false) && r.advanceChar(' ') && r.advanceYear() {
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// RFC850, e.g.
+		//  - "Monday, 02-Jan-06 15:04:05 MST"
+		backtrackedSlice := r.source[r.head-3:]
+		var matchedDay []byte
+		for _, dw := range longDayNames {
+			if bytes.HasPrefix(backtrackedSlice, dw) {
+				matchedDay = dw
+				break
+			}
+		}
+		if matchedDay != nil {
+			r.head += len(matchedDay) - 3
+			if r.advanceChar(',') && r.advanceChar(' ') && r.advanceUintUpTo(31, 2) && r.advanceChar('-') && r.advanceMonthName() && r.advanceChar('-') && r.advanceUintUpTo(99, 2) && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(true) {
+				r.emit(false, placeholderTimestamp)
+				return true
+			}
+		}
+
+		r.cur -= 3 // unconsume
+		r.resetHead()
+		return false
+	}
+	r.resetHead()
+
+	if r.advanceMonthName() {
+		r.commit() // provisionally consume this
+
+		// Linux journald logs and others similar like this:
+		//  - Feb 29 23:00:14
+		//  - Apr-10 23:43:46.807
+		//  - Jul  1 00:21:28
+		if (r.advanceChar('-') || (r.advanceChar(' ') && (r.advanceChar(' ') || true))) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceTime(false) {
+			r.commit()
+			// This is already a timestamp, but let's try to match subseconds as well
+			if r.advanceChar('.') && r.advanceUint() {
+				r.commit()
+			} else {
+				r.resetHead()
+			}
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.cur -= 3 // unconsume
+		r.resetHead()
+		return false
+	}
+	r.resetHead()
+
+	return false
+}
+
+func (r *replacer) wasNumPreservingKeyword() bool {
+	for _, key := range numPreservingKeys {
+		pPos := r.cur - 1 - len(key)
+		if pPos < -1 {
+			return false // all subsequent keys are longer
+		}
+		if pPos != -1 && !boundaryChars[r.source[pPos]] {
+			continue
+		}
+		if bytes.HasPrefix(r.source[pPos+1:], key) {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *replacer) replaceWithPlaceholders() {
+	lineLen := len(r.source)
+
+	var c byte
+	onBoundary := true
+	for r.cur = 0; r.cur < lineLen; r.cur++ {
+		c = r.source[r.cur]
+		switch {
+		// If we're currently not at a boundary, the only thing we need to check
+		// is whether the current char would create a boundary in the next iteration.
+		case !onBoundary:
+			onBoundary = boundaryChars[c]
+
+			// We weren't at a boundary and now we are, so check if we match one
+			// of the special keywords that will preserve numbers
+			if onBoundary {
```

Comment:
Hmm :thinking: Maybe, though I am not sure if this will always hold. For example, the `_` in `status_code=` is technically a boundary character, so it's not inconceivable for `status` to already be "flushed" when we reach the `=`.

Reply:
Yeah, just stepped through the `status_code=123` example with a debugger and that's exactly what happens, because `_` is a boundary and `c` is a potential start of a hex string, we encounter this branch and copy `status_` to the `r.dest` slice (so `r.cur - r.tail` becomes 0) :disappointed: 

Let's keep this thread open though, as a potential performance improvement along with the other ones I opened, because that's an implementation detail that's fixable. If profiling proves that this would make a big difference, we can change the other behavior to be less aggressive in copying the source string